controlplane/kubeadm/internal/controllers/status.go

```diff
@@ -135,3 +142,257 @@ func (r *KubeadmControlPlaneReconciler) updateStatus(ctx context.Context, contro
+		Message: strings.Join(messages, "; "),
+	})
+}
+
+func setMachinesReadyCondition(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane, machines collections.Machines) {
+	if len(machines) == 0 {
+		v1beta2conditions.Set(kcp, metav1.Condition{
+			Type:   controlplanev1.KubeadmControlPlaneMachinesReadyV1Beta2Condition,
+			Status: metav1.ConditionTrue,
+			Reason: controlplanev1.KubeadmControlPlaneMachinesReadyNoReplicasV1Beta2Reason,
+		})
+		return
+	}
+
+	readyCondition, err := v1beta2conditions.NewAggregateCondition(
+		machines.UnsortedList(), clusterv1.MachineReadyV1Beta2Condition,
+		v1beta2conditions.TargetConditionType(controlplanev1.KubeadmControlPlaneMachinesReadyV1Beta2Condition),
+	)
+	if err != nil {
+		v1beta2conditions.Set(kcp, metav1.Condition{
+			Type:    controlplanev1.KubeadmControlPlaneMachinesReadyV1Beta2Condition,
+			Status:  metav1.ConditionUnknown,
+			Reason:  controlplanev1.KubeadmControlPlaneMachinesReadyInternalErrorV1Beta2Reason,
+			Message: "Please check controller logs for errors",
+		})
+
+		log := ctrl.LoggerFrom(ctx)
+		log.Error(err, fmt.Sprintf("Failed to compute %s condition", controlplanev1.KubeadmControlPlaneMachinesReadyV1Beta2Condition))
+		return
+	}
+
+	v1beta2conditions.Set(kcp, *readyCondition)
+}
+
+func setMachinesUpToDateCondition(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane, machines collections.Machines) {
+	if len(machines) == 0 {
+		v1beta2conditions.Set(kcp, metav1.Condition{
+			Type:   controlplanev1.KubeadmControlPlaneMachinesUpToDateV1Beta2Condition,
+			Status: metav1.ConditionTrue,
+			Reason: controlplanev1.KubeadmControlPlaneMachinesUpToDateNoReplicasV1Beta2Reason,
+		})
+		return
+	}
+
+	upToDateCondition, err := v1beta2conditions.NewAggregateCondition(
+		machines.UnsortedList(), clusterv1.MachineUpToDateV1Beta2Condition,
+		v1beta2conditions.TargetConditionType(controlplanev1.KubeadmControlPlaneMachinesUpToDateV1Beta2Condition),
+	)
+	if err != nil {
+		v1beta2conditions.Set(kcp, metav1.Condition{
+			Type:    controlplanev1.KubeadmControlPlaneMachinesUpToDateV1Beta2Condition,
+			Status:  metav1.ConditionUnknown,
+			Reason:  controlplanev1.KubeadmControlPlaneMachinesUpToDateInternalErrorV1Beta2Reason,
+			Message: "Please check controller logs for errors",
+		})
+
+		log := ctrl.LoggerFrom(ctx)
+		log.Error(err, fmt.Sprintf("Failed to compute %s condition", controlplanev1.KubeadmControlPlaneMachinesUpToDateV1Beta2Condition))
+		return
+	}
+
+	v1beta2conditions.Set(kcp, *upToDateCondition)
+}
+
+func calculateMissingReferencesMessage(kcp *controlplanev1.KubeadmControlPlane, infraMachineTemplateNotFound bool) string {
+	if infraMachineTemplateNotFound {
+		return fmt.Sprintf("because %s does not exist", kcp.Spec.MachineTemplate.InfrastructureRef.Kind)
+	}
+	return ""
+}
+
+func setRemediatingCondition(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane, unhealthyMachines collections.Machines) {
+	if len(unhealthyMachines) == 0 {
```

Comment:
Okay I'm a bit confused. Here we look for unhealthyMachines, while below we aggregate based on ownerRemedated. I would have expected both to look for ownerRemedaited (maybe let's just talk about it)

Reply:
Fixed as per discussion (the condition must focus on what KCP is responsible for)